File: packages/sdk-py/src/agenttool/traces.py
"""Traces client for the agent-trace reasoning provenance API."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, TYPE_CHECKING, TypedDict, Union

import httpx

from .exceptions import AgentToolError

if TYPE_CHECKING:
    pass


class TraceAlternative(TypedDict):
    """One structured alternative and its caller-supplied rejection reason."""

    option: str
    why_not: str


TraceAlternativeValue = Union[str, TraceAlternative]


@dataclass
class Trace:
    """A stored reasoning trace."""

    id: str
    trace_id: str
    # Kept third for positional constructor compatibility. Live routes omit it.
    project_id: Optional[str]
    decision_type: str
    decision_summary: str
    observations: List[str]
    conclusion: str
    created_at: str
    agent_id: Optional[str] = None
    identity_id: Optional[str] = None
    session_id: Optional[str] = None
    output_ref: Optional[str] = None
    hypothesis: Optional[str] = None
    confidence: Optional[float] = None
    alternatives: Optional[List[TraceAlternativeValue]] = None
    signals: Optional[Dict[str, Any]] = None
    files_read: Optional[List[str]] = None
    key_facts: Optional[List[str]] = None
    external_signals: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None
    parent_trace_id: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    signature: Optional[str] = None
    signing_key_id: Optional[str] = None
    has_signature: Optional[bool] = None


@dataclass
class TraceSearchResult:
    """A trace search result with its Postgres full-text rank score."""

    trace: Trace
    score: float


@dataclass
class TraceChain:
    """A reasoning lineage with aliases for the former SDK shape."""

    root: Trace
    ancestors: List[Trace] = field(default_factory=list)
    descendants: List[Trace] = field(default_factory=list)
    counts: Dict[str, int] = field(
        default_factory=lambda: {"ancestors": 0, "descendants": 0}
    )
    _legacy_depth: Optional[int] = field(default=None, repr=False)

    @property
    def parent(self) -> Trace:
        """Deprecated alias for root."""
        return self.root

    @property
    def children(self) -> List[Trace]:
        """Deprecated alias for descendants."""
        return self.descendants

    @property
    def depth(self) -> int:
        """Deprecated alias for the number of ancestors."""
        if self._legacy_depth is not None:
            return self._legacy_depth
        return len(self.ancestors)


def _dict_to_trace(d: Dict[str, Any]) -> Trace:
    return Trace(
        id=d["id"],
        trace_id=d["trace_id"],
        decision_type=d["decision_type"],
        decision_summary=d["decision_summary"],
        observations=d.get("observations") or [],
        conclusion=d["conclusion"],
        created_at=d.get("created_at", ""),
        project_id=d.get("project_id"),
        agent_id=d.get("agent_id"),
        identity_id=d.get("identity_id"),
        session_id=d.get("session_id"),
        output_ref=d.get("output_ref"),
        hypothesis=d.get("hypothesis"),
        confidence=d.get("confidence"),
        alternatives=d.get("alternatives"),
        signals=d.get("signals"),
        files_read=d.get("files_read"),
        key_facts=d.get("key_facts"),
        external_signals=d.get("external_signals"),
        tags=d.get("tags"),
        parent_trace_id=d.get("parent_trace_id"),
        metadata=d.get("metadata"),
        signature=d.get("signature"),
        signing_key_id=d.get("signing_key_id"),
        has_signature=d.get("has_signature"),
    )


class TracesClient:
    """Client for the agent-trace reasoning provenance API.

    Usage::

        at = AgentTool()

        # Store a reasoning trace
        trace = at.traces.store(
            observations=["User asked which capability handles X", "Searched marketplace listings"],
            conclusion="Listing 'capability/Y' matches; price acceptable; invoke",
            decision_type="decision",
            tags=["marketplace", "invocation"],
        )

        # Search traces with Postgres full-text search
        results = at.traces.search("billing decisions", limit=5)

        # Retrieve a specific trace
        t = at.traces.get(trace.trace_id)
    """

    def __init__(self, http: httpx.Client, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    def _url(self, path: str) -> str:
        return f"{self._base}{path}"

    def _raise(self, resp: httpx.Response) -> None:
        try:
            payload: Any = resp.json()
        except Exception:
            payload = None
        raise AgentToolError.from_response_body(
            payload,
            resp.status_code,
            fallback=f"Traces API request failed ({resp.status_code}).",
            headers=resp.headers,
        )

    def store(
        self,
        *,
        observations: List[str],
        conclusion: str,
        decision_type: str = "decision",
        decision_summary: Optional[str] = None,
        agent_id: Optional[str] = None,
        identity_id: Optional[str] = None,
        session_id: Optional[str] = None,
        output_ref: Optional[str] = None,
        hypothesis: Optional[str] = None,
        confidence: Optional[float] = None,
        alternatives: Optional[List[TraceAlternativeValue]] = None,
        signals: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None,
        parent_trace_id: Optional[str] = None,
        files_read: Optional[List[str]] = None,
        key_facts: Optional[List[str]] = None,
        external_signals: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Trace:
        """Store a reasoning trace.

        Args:
            observations: List of observation strings that led to the decision.
            conclusion: What was concluded or decided.
            decision_type: One of: tool_call, memory_write, plan, decision, verification, other.
            decision_summary: Short summary (defaults to first 120 chars of conclusion).
            agent_id: Optional agent identifier.
            identity_id: Optional AgentTool identity UUID.
            session_id: Optional session identifier.
            output_ref: Optional reference to the output this trace explains.
            hypothesis: Optional hypothesis considered.
            confidence: Optional 0.0–1.0 confidence score.
            alternatives: Optional structured alternatives. Legacy strings become
                entries with an empty why_not; the SDK never invents a reason.
            signals: Optional structured signals used in the reasoning itself.
            tags: Optional list of tags for classification.
            parent_trace_id: Optional parent trace_id for chaining.
            files_read: Optional list of files read during reasoning.
            key_facts: Optional key facts extracted during reasoning.
            external_signals: Optional namespaced reports produced outside AgentTool.
                Passing these uploads server-readable trace context; the SDK never
                analyzes or transmits a report implicitly.
            metadata: Optional caller metadata. The API overwrites client_source
                with a best-effort origin label, not an attestation or security
                boundary.

        Returns:
            The created Trace object.
        """
        # API expects nested decision + reasoning objects
        decision: Dict[str, Any] = {
            "type": decision_type,
            "summary": decision_summary or conclusion[:120],
        }
        if output_ref is not None:
            decision["output_ref"] = output_ref

        reasoning: Dict[str, Any] = {
            "observations": observations,
            "conclusion": conclusion,
        }
        if hypothesis is not None:
            reasoning["hypothesis"] = hypothesis
        if confidence is not None:
            reasoning["confidence"] = confidence
        if alternatives is not None:
            reasoning["alternatives"] = self._normalize_alternatives(alternatives)
        if signals is not None:
            reasoning["signals"] = signals

        body: Dict[str, Any] = {"decision": decision, "reasoning": reasoning}
        for k, v in [
            ("agent_id", agent_id),
            ("identity_id", identity_id),
            ("session_id", session_id),
            ("tags", tags),
            ("parent_trace_id", parent_trace_id),
            ("metadata", metadata),
        ]:
            if v is not None:
                body[k] = v

        context: Dict[str, Any] = {}
        if files_read is not None:
            context["files_read"] = files_read
        if key_facts is not None:
            context["key_facts"] = key_facts
        if external_signals is not None:
            context["external_signals"] = external_signals
        if context:
            body["context"] = context

        resp = self._http.post(self._url("/v1/traces"), json=body)
        if resp.is_error:
            self._raise(resp)

        created = resp.json()
        return self.get(created["trace_id"])

    def get(self, trace_id: str) -> Trace:
        """Retrieve a trace by its trace_id.

        Args:
            trace_id: The trace_id returned by store().

        Returns:
            The Trace object.
        """
        resp = self._http.get(self._url(f"/v1/traces/{trace_id}"))
        if resp.is_error:
            self._raise(resp)
        return _dict_to_trace(resp.json())

    def search(
        self,
        query: str,
        *,
        limit: int = 10,
        agent_id: Optional[str] = None,
        identity_id: Optional[str] = None,
        session_id: Optional[str] = None,
        decision_type: Optional[str] = None,
        tag: Optional[str] = None,
    ) -> List[TraceSearchResult]:
        """Search traces using the API's Postgres full-text index.

        Args:
            query: Natural language search query.
            limit: Maximum number of results (default 10).
            agent_id: Filter by agent_id.
            identity_id: Filter by identity_id.
            session_id: Filter by session_id.
            decision_type: Filter by decision type.
            tag: Deprecated compatibility option. The live route has no tag
                filter, so this value is intentionally ignored.

        Returns:
            Ranked list of TraceSearchResult objects with Postgres rank scores.
        """
        body: Dict[str, Any] = {"query": query, "limit": limit}
        if agent_id is not None:
            body["agent_id"] = agent_id
        if identity_id is not None:
            body["identity_id"] = identity_id
        if session_id is not None:
            body["session_id"] = session_id
        if decision_type is not None:
            body["decision_type"] = decision_type
        # `tag` stays accepted for source compatibility, but the live search
        # schema has no tag filter and silently stripped this field historically.

        resp = self._http.post(self._url("/v1/traces/search"), json=body)
        if resp.is_error:
            self._raise(resp)

        payload = resp.json()
        rows = payload if isinstance(payload, list) else payload.get("results", [])
        results: List[TraceSearchResult] = []
        for row in rows:
            if "trace" in row:
                results.append(
                    TraceSearchResult(
                        trace=_dict_to_trace(row["trace"]), score=float(row["score"])
                    )
                )
            else:
                results.append(
                    TraceSearchResult(trace=_dict_to_trace(row), score=float(row["score"]))
                )
        return results

    def chain(self, trace_id: str) -> TraceChain:
        """Retrieve the reasoning lineage for a trace.

        Args:
            trace_id: The parent trace_id.

        Returns:
            TraceChain with root, ancestors, descendants, counts, and legacy aliases.
        """
        resp = self._http.get(self._url(f"/v1/traces/chain/{trace_id}"))
        if resp.is_error:
            self._raise(resp)

        data = resp.json()
        root_data = data.get("root") or data["parent"]
        ancestors_data = data.get("ancestors", [])
        descendants_data = data.get("descendants", data.get("children", []))
        return TraceChain(
            root=_dict_to_trace(root_data),
            ancestors=[_dict_to_trace(t) for t in ancestors_data],
            descendants=[_dict_to_trace(t) for t in descendants_data],
            counts=dict(
                data.get("counts")
                or {
                    "ancestors": len(ancestors_data),
                    "descendants": len(descendants_data),
                }
            ),
            _legacy_depth=data.get("depth"),
        )

    def delete(self, trace_id: str) -> None:
        """Delete a trace.

        Args:
            trace_id: The trace_id to delete.
        """
        resp = self._http.delete(self._url(f"/v1/traces/{trace_id}"))
        if resp.is_error:
            self._raise(resp)

    def _normalize_alternatives(
        self, alternatives: List[TraceAlternativeValue]
    ) -> List[TraceAlternative]:
        normalized: List[TraceAlternative] = []
        for index, alternative in enumerate(alternatives, start=1):
            if isinstance(alternative, str):
                normalized.append({"option": alternative, "why_not": ""})
                continue
            if (
                not isinstance(alternative, dict)
                or not isinstance(alternative.get("option"), str)
                or not isinstance(alternative.get("why_not"), str)
            ):
                raise AgentToolError(
                    f"Trace alternative {index} needs both option and why_not strings.",
                    hint=(
                        "Use alternatives=[{'option': '...', 'why_not': '...'}] "
                        "or legacy string entries. The SDK will not invent a "
                        "rejection reason."
                    ),
                )
            normalized.append(
                {"option": alternative["option"], "why_not": alternative["why_not"]}
            )
        return normalized

File: packages/sdk-ts/src/traces.ts
/**
 * Traces client for the agent-trace reasoning provenance API.
 */

import { AgentToolError } from "./errors.js";
import type { HttpConfig } from "./_http.js";

/** One structured alternative considered while reaching a conclusion. */
export interface TraceAlternative {
  option: string;
  /** Why it was not chosen. The SDK never invents this reason. */
  why_not: string;
}

/** Legacy string alternatives remain readable and accepted at the SDK edge. */
export type TraceAlternativeValue = TraceAlternative | string;

/** A stored reasoning trace. */
export interface Trace {
  id: string;
  trace_id: string;
  agent_id?: string | null;
  identity_id?: string | null;
  /** Present only in legacy responses; current trace routes do not expose it. */
  project_id?: string;
  session_id?: string | null;
  created_at: string;
  decision_type: string;
  decision_summary: string;
  output_ref?: string | null;
  observations: string[];
  hypothesis?: string | null;
  conclusion: string;
  confidence?: number | null;
  alternatives?: TraceAlternativeValue[] | null;
  signals?: Record<string, unknown> | null;
  files_read?: string[] | null;
  key_facts?: string[] | null;
  external_signals?: Record<string, unknown> | null;
  tags?: string[] | null;
  parent_trace_id?: string | null;
  metadata?: Record<string, unknown>;
  signature?: string | null;
  signing_key_id?: string | null;
  has_signature?: boolean;
}

/** Options for storing a trace. */
export interface StoreTraceOptions {
  /** Free-form observation strings that led to the decision. */
  observations: string[];
  /** What was concluded / decided. */
  conclusion: string;
  /** One of: tool_call | memory_write | plan | decision | verification | other */
  decision_type?: string;
  /** Short human-readable summary of the decision. */
  decision_summary?: string;
  agent_id?: string;
  identity_id?: string;
  session_id?: string;
  output_ref?: string;
  hypothesis?: string;
  confidence?: number;
  /**
   * Structured alternatives, or legacy strings. A legacy string is sent as
   * `{ option, why_not: "" }`, preserving the absence of a supplied reason.
   */
  alternatives?: TraceAlternativeValue[];
  /** Structured signals used in the reasoning itself. */
  signals?: Record<string, unknown>;
  tags?: string[];
  parent_trace_id?: string;
  files_read?: string[];
  key_facts?: string[];
  /**
   * Namespaced reports produced outside AgentTool, such as an explicitly
   * supplied local RhetorLint signal. Passing this uploads server-readable
   * trace context; the SDK never analyzes or transmits a report implicitly.
   */
  external_signals?: Record<string, unknown>;
  /**
   * Caller metadata. The API overwrites client_source with a best-effort
   * origin label; that soft signal is not an attestation or security boundary.
   */
  metadata?: Record<string, unknown>;
}

/** A search result entry. */
export interface TraceSearchResult {
  trace: Trace;
  score: number;
}

/** Options for searching traces. */
export interface SearchTracesOptions {
  /** Maximum number of results (default 10). */
  limit?: number;
  /** Filter by agent_id. */
  agent_id?: string;
  /** Filter by session_id. */
  session_id?: string;
  /** Filter by identity_id. */
  identity_id?: string;
  /** Filter by decision type. */
  decision_type?: string;
  /**
   * @deprecated The live full-text search route has no tag filter. Retained as
   * an ignored option so existing callers keep compiling.
   */
  tag?: string;
}

/** A reasoning lineage, with compatibility aliases for the former SDK shape. */
export interface TraceChain {
  root: Trace;
  ancestors: Trace[];
  descendants: Trace[];
  counts: {
    ancestors: number;
    descendants: number;
  };
  /** @deprecated Alias for root. */
  parent: Trace;
  /** @deprecated Alias for descendants. */
  children: Trace[];
  /** @deprecated Alias for ancestors.length. */
  depth: number;
}

/**
 * Client for the agent-trace reasoning provenance API.
 *
 * @example
 * ```ts
 * const at = new AgentTool();
 *
 * // Store a reasoning trace
 * const trace = await at.traces.store({
 *   observations: ["User asked about pricing", "Checked tier table"],
 *   conclusion: "User is on Free tier, eligible to upgrade",
 *   decision_type: "decision",
 *   tags: ["billing", "upgrade"],
 * });
 *
 * // Search traces with Postgres full-text search
 * const results = await at.traces.search("billing decisions", { limit: 5 });
 *
 * // Retrieve a specific trace
 * const t = await at.traces.get(trace.trace_id);
 * ```
 */
export class TracesClient {
  private readonly http: HttpConfig;

  /** @internal */
  constructor(http: HttpConfig) {
    this.http = http;
  }

  /**
   * Store a reasoning trace.
   *
   * @param options - Trace content (observations + conclusion required).
   * @returns The created Trace object with its trace_id.
   */
  async store(options: StoreTraceOptions): Promise<Trace> {
    const decision: Record<string, unknown> = {
      type: options.decision_type ?? "decision",
      summary: options.decision_summary || options.conclusion.slice(0, 120),
    };
    if (options.output_ref !== undefined) decision.output_ref = options.output_ref;

    const reasoning: Record<string, unknown> = {
      observations: options.observations,
      conclusion: options.conclusion,
    };
    if (options.hypothesis !== undefined) reasoning.hypothesis = options.hypothesis;
    if (options.confidence !== undefined) reasoning.confidence = options.confidence;
    if (options.alternatives !== undefined) {
      reasoning.alternatives = this._normalizeAlternatives(options.alternatives);
    }
    if (options.signals !== undefined) reasoning.signals = options.signals;

    const context: Record<string, unknown> = {};
    if (options.files_read !== undefined) context.files_read = options.files_read;
    if (options.key_facts !== undefined) context.key_facts = options.key_facts;
    if (options.external_signals !== undefined) {
      context.external_signals = options.external_signals;
    }

    const body: Record<string, unknown> = { decision, reasoning };
    if (options.agent_id !== undefined) body.agent_id = options.agent_id;
    if (options.identity_id !== undefined) body.identity_id = options.identity_id;
    if (options.session_id !== undefined) body.session_id = options.session_id;
    if (options.tags !== undefined) body.tags = options.tags;
    if (options.parent_trace_id !== undefined) body.parent_trace_id = options.parent_trace_id;
    if (options.metadata !== undefined) body.metadata = options.metadata;
    if (Object.keys(context).length > 0) body.context = context;

    const resp = await this.http.request(`${this.http.baseUrl}/v1/traces`, {
      method: "POST",
      headers: this.http.headers,
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(this.http.timeout),
    });

    if (resp.status >= 400) {
      throw await this._responseError(resp);
    }

    const created = (await resp.json()) as { trace_id: string };
    // Return the full trace by fetching it
    return this.get(created.trace_id);
  }

  /**
   * Retrieve a trace by its trace_id.
   *
   * @param traceId - The trace_id returned by store().
   */
  async get(traceId: string): Promise<Trace> {
    const resp = await this.http.request(`${this.http.baseUrl}/v1/traces/${traceId}`, {
      headers: this.http.headers,
      signal: AbortSignal.timeout(this.http.timeout),
    });

    if (resp.status >= 400) {
      throw await this._responseError(resp);
    }

    return (await resp.json()) as Trace;
  }

  /**
   * Search traces using the API's Postgres full-text index.
   *
   * @param query - Natural language query.
   * @param options - Filters: limit, agent_id, identity_id, session_id, and
   * decision_type. The legacy tag option is accepted but ignored because the
   * live route does not support it.
   * @returns Ranked list of matching traces with Postgres ts_rank scores.
   */
  async search(query: string, options?: SearchTracesOptions): Promise<TraceSearchResult[]> {
    const body: Record<string, unknown> = {
      query,
      limit: options?.limit ?? 10,
    };
    if (options?.agent_id !== undefined) body.agent_id = options.agent_id;
    if (options?.identity_id !== undefined) body.identity_id = options.identity_id;
    if (options?.session_id !== undefined) body.session_id = options.session_id;
    if (options?.decision_type !== undefined) body.decision_type = options.decision_type;

    const resp = await this.http.request(`${this.http.baseUrl}/v1/traces/search`, {
      method: "POST",
      headers: this.http.headers,
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(this.http.timeout),
    });

    if (resp.status >= 400) {
      throw await this._responseError(resp);
    }

    const payload = (await resp.json()) as unknown;
    const rows = Array.isArray(payload)
      ? payload
      : ((payload as { results?: unknown[] } | null)?.results ?? []);

    return rows.map((entry) => {
      const row = entry as Record<string, unknown>;
      // Tolerate the former SDK-shaped response while preferring the live
      // route's flat `Trace & { score }` rows.
      if (row.trace && typeof row.trace === "object") {
        return { trace: row.trace as Trace, score: Number(row.score) };
      }
      const { score, ...trace } = row;
      return { trace: trace as unknown as Trace, score: Number(score) };
    });
  }

  /**
   * Retrieve the reasoning lineage for a trace.
   *
   * @param traceId - The parent trace_id.
   */
  async chain(traceId: string): Promise<TraceChain> {
    const resp = await this.http.request(`${this.http.baseUrl}/v1/traces/chain/${traceId}`, {
      headers: this.http.headers,
      signal: AbortSignal.timeout(this.http.timeout),
    });

    if (resp.status >= 400) {
      throw await this._responseError(resp);
    }

    const data = (await resp.json()) as Partial<TraceChain>;
    const root = data.root ?? data.parent!;
    const ancestors = data.ancestors ?? [];
    const descendants = data.descendants ?? data.children ?? [];
    const counts = data.counts ?? {
      ancestors: ancestors.length,
      descendants: descendants.length,
    };
    return {
      root,
      ancestors,
      descendants,
      counts,
      parent: root,
      children: descendants,
      depth: data.depth ?? ancestors.length,
    };
  }

  /**
   * Delete a trace.
   *
   * @param traceId - The trace_id to delete.
   */
  async delete(traceId: string): Promise<void> {
    const resp = await this.http.request(`${this.http.baseUrl}/v1/traces/${traceId}`, {
      method: "DELETE",
      headers: this.http.headers,
      signal: AbortSignal.timeout(this.http.timeout),
    });

    if (resp.status >= 400) {
      throw await this._responseError(resp);
    }
  }

  private async _responseError(resp: Response): Promise<AgentToolError> {
    let payload: unknown;
    try {
      payload = await resp.json();
    } catch {
      payload = undefined;
    }
    return AgentToolError.fromResponseBody(
      payload,
      resp.status,
      `Traces API request failed (${resp.status}).`,
      resp.headers,
    );
  }

  private _normalizeAlternatives(
    alternatives: TraceAlternativeValue[],
  ): TraceAlternative[] {
    return alternatives.map((alternative, index) => {
      if (typeof alternative === "string") {
        return { option: alternative, why_not: "" };
      }
      if (
        typeof alternative !== "object" ||
        alternative === null ||
        typeof alternative.option !== "string" ||
        typeof alternative.why_not !== "string"
      ) {
        throw new AgentToolError(
          `Trace alternative ${index + 1} needs both option and why_not strings.`,
          {
            hint:
              'Use alternatives: [{ option: "...", why_not: "..." }] or legacy string entries. The SDK will not invent a rejection reason.',
          },
        );
      }
      return { option: alternative.option, why_not: alternative.why_not };
    });
  }
}
